app/models/operatingsystem.rb

```diff
@@ -255,8 +255,14 @@ def boot_filename(host = nil)
     return default_boot_filename if host.nil? || host.pxe_loader.nil?
     return host.foreman_url('iPXE') if host.pxe_loader == 'iPXE Embedded'
     architecture = host.arch.nil? ? '' : host.arch.bootfilename_efi
-    boot_uri = URI.parse((host.subnet&.httpboot?) ? host.subnet.httpboot.url : Setting[:unattended_url])
-    self.class.all_loaders_map(architecture, "#{boot_uri.host}:#{boot_uri.port}")[host.pxe_loader]
+    if host.subnet&.httpboot?
+      port = host.subnet.httpboot.setting(:HTTPBoot, 'http_port') || 80
```

Comment:
Why port 80? Do you assume Foreman is running on the same host or did you intend `8000`?

It is possible that `http_port` is set but the module is https-only. That means you should also check the `http_enabled` setting. Perhaps we should amend the smart-proxy side to ensure the port is nil unless http_enabled is true?

Reply:
> Why port 80?

My thinking is that user wants to run her own HTTP server, which is usually 80. Another option is to raise an error and insist on setting being exposed which probably makes more sense and creates less confusion.

> It is possible that http_port is set but the module is https-only.

That is not possible, I tested this. In that case `http_port` is `nil`.

Reply:
On the other hand, defaulting to `8000` enables use with older smart proxies.

Reply:
> > It is possible that http_port is set but the module is https-only.
> 
> That is not possible, I tested this. In that case `http_port` is `nil`

It is. Reversed, but I tested this in the proxy by writing an integration test:

```
def test_features_http
  Proxy::DefaultModuleLoader.any_instance.expects(:load_configuration_file).with('httpboot.yml').returns(enabled: 'http', root_dir: '/var/lib/tftpboot')

  get '/features'

  response = JSON.parse(last_response.body)

  mod = response['httpboot']
  refute_nil(mod)
  assert(mod['http_enabled'])
  refute(mod['https_enabled'])
  assert_equal('running', mod['state'], Proxy::LogBuffer::Buffer.instance.info[:failed_modules][:httpboot])
  assert_equal([], mod['capabilities'])

  expected_settings = {'http_port' => nil, 'https_port' => 8443}
  assert_equal(expected_settings, mod['settings'])
end
```

Note `https_port` is set, but `https_enabled` is false because the module is configured with`enabled: 'http'`. That was what I was thinking. IMHO we should fix this on the proxy side so the `*_port` values are nil when it's not enabled so you don't have to think about this on the Foreman side.

> On the other hand, defaulting to `8000` enables use with older smart proxies.

I think this is more logical. We don't really support custom HTTP servers.

Reply:
I was confused by this:

```ruby
    def http_enabled?
      !settings.http_port.nil?
    end
```

I thought it's the same setting and haven't considered that whole httpboot module can be enabled via `enabled: 'http'`. I agree, let me change this on the proxy side then.

Reply:
Currently we can't distinguish between a proxy that doesn't have the settings exposed (too old) and doesn't have it enabled. For that we need to check if the key is set. I'm slightly leaning to not maintaining compatibility with older proxies since HTTP boot is a relative new feature. Then we show a clear error to the user. Do you agree we should drop the defaults on the Foreman side?

Reply:
https://github.com/theforeman/smart-proxy/pull/677